src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -279,38 +282,51 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
             {
                 return null;
             }
-
             var headers = httpContext.Request.Headers;
-            var requestId = headers.TraceParent;
-            if (requestId.Count == 0)
-            {
-                requestId = headers.RequestId;
-            }
+            _propagator.ExtractTraceIdAndState(headers,
+                static (object? carrier, string fieldName, out string? fieldValue, out IEnumerable<string>? fieldValues) =>
+                {
+                    fieldValue = default;
+                    fieldValues = default;
+                    if (carrier is null)
+                    {
+                        return;
+                    }
+                    var headers = (IHeaderDictionary)carrier;
+                    fieldValue = headers[fieldName];
+                },
+                out var requestId,
+                out var traceState);
 
-            if (!StringValues.IsNullOrEmpty(requestId))
+            if (requestId is not null)
```

Comment:
>requestId is not null) [](http://example.com/codeflow?start=16&length=22)

does 

fieldValue = headers[fieldName];

return null or empty strings for non-found headers? is it possible it throws?

I am asking because we are checking only for null here.

Reply:
Ahh yes. I didn't realize empty value HTTP headers are legal. Will update this